### DISCUSSION TITLE
In webhooks, how do we handle different endpoints based on different event types?

### DISCUSSION THREAD

**Discussion by ztzsws1** [elsa-workflows/elsa-core#6238] [Design & tech]
Hello @sfmskywalker , I have multiple EventTypes, each corresponding to multiple Endpoint addresses. I know how to configure them in JSON. 
```json
{
  "Webhooks": {
    "Endpoints": [
      {
        "EventTypes": ["RunTask"],
        "Url": "https://localhost:5002/api/webhooks/run-task"
      },
      {
        "EventTypes": ["AnotherTask"],
        "Url": "https://localhost:5002/api/webhooks/another-task"
      }
      // Add more webhook endpoints as needed
    ]
  }
}
```

However, I've found that in the RunTaskHandler, only the EventType of "RunTask" can be executed, as shown in the figure. 

![webhooks](https://github.com/user-attachments/assets/d57e06fc-d171-47b5-a37d-7ed29e344d2f)

I'm wondering if you have any suggestions for my scenario? Do I need to write a custom RunTaskHandler? Can we add an additional EventType parameter to RunTask? The default value could be RunTask. Looking forward to your reply. 

Best Regards

**Comment by EngSayed** [comment]
I am still new to Elsa but my understanding is that you write your own Task Handler which Elsa Server will call for that task event type.
For example in the External App Integration in the [documentation](https://v3.elsaworkflows.io/docs/guides/external-app-integration)
You have an external application which has `RunTask` endpoint mapped to `api/webhooks/run-task`:
```
[ApiController]
[Route("api/webhooks")]
public class WebhookController(OnboardingDbContext dbContext) : Controller
{
    [HttpPost("run-task")]
    public async Task<IActionResult> RunTask(WebhookEvent webhookEvent)
    {
        var payload = webhookEvent.Payload;
        var taskPayload = payload.TaskPayload;
        var employee = taskPayload.Employee;
        string eventType = webhookEvent.EventType.ToString();

        var task = new OnboardingTask
        {
            ProcessId = payload.WorkflowInstanceId,
            ExternalId = payload.TaskId,
            Name = payload.TaskName,
            Description = taskPayload.Description,
            EmployeeEmail = employee.Email,
            EmployeeName = employee.Name,
            CreatedAt = DateTimeOffset.Now
        };

        await dbContext.Tasks.AddAsync(task);
        await dbContext.SaveChangesAsync();

        return Ok();
    }
}
```

So you will need to have a different endpoint for the second Event Type and so on.

**Comment by sfmskywalker** [comment]
The configured endpoint routing is based on the event type that Elsa publishes. For the built-in **Run Task** notification, that type is fixed: the handler emits `Elsa.RunTask`; it does not expose a setting to change that per task.

For another event type, use a separate/custom notification handler that publishes the desired event type and configure the corresponding endpoint for it. The external application can then handle that endpoint independently, as in the external-app integration guide.

The current implementation still constructs `NewWebhookEvent("Elsa.RunTask", ...)` in [`RunTaskHandler`](https://github.com/elsa-workflows/elsa-core/blob/main/src/modules/Elsa.Http.Webhooks/Handlers/RunTaskHandler.cs#L16-L40), so adding another endpoint alone will not cause a different event type to be emitted.

Closing this as answered.
